python/ray/rllib/ppo/loss.py

```diff
@@ -93,3 +94,115 @@ def compute(self, observation):
 
     def loss(self):
         return self.loss
+
+class mProximalPolicyLoss:
+    other_output = ["vf_preds", "logprobs"]
+    is_recurrent = False
+
+    def __init__(
+            self, env,
+            observations, value_targets, advantages, actions,
+            prev_logits, prev_vf_preds, logit_dim,
+            kl_coeff, distribution_class, config, sess):
+
+        assert (isinstance(env.action_space, gym.spaces.Discrete) or
+                isinstance(env.action_space, gym.spaces.Box))
+
+        # store a copy of the observations for the feed dict
+        self.observations = observations
+        split_observations = env.split_input_tensor(observations)
+        act_list = env.split_output_tensor(actions)
+        prev_logits = env.split_output_tensor(prev_logits)
+        split_advantages = env.split_along_agents(advantages)
+        split_logit_dim = env.split_output_number(logit_dim)
+
+        self.prev_dist = []
+        self.curr_logits = []
+        self.curr_dist = []
+        self.sampler = []
+        self.ratio = []
+        self.kl = []
+        self.mean_kl = []
+        self.entropy = []
+        self.mean_entropy = []
+        self.surr1 = []
+        self.surr2 = []
+        self.surr = []
+        self.mean_policy_loss = []
+        for i in range(len(act_list)):
+            config["model_num"] = str(i)
+            self.prev_dist.append(distribution_class(prev_logits[i]))
+            self.curr_logits.append(ModelCatalog.get_model(
```

Comment:
I think you can actually use the original policy loss and instead use a custom model that can interpret the N-agent observations of shape `[N, original_shape]` and output `[N, action_space]`. GAE might work with no special handling then.

Reply:
@ericl Sorry, was on break all week. This seems like a sensible approach, and we can just extend GAE later if we need non-shared rewards?

Reply:
As an update, this seems to be working. I'll push up a cleaned up version soon.

Reply:
Awesome, looking forward to it.

For the non-shared reward case: yeah, we'd need something like that, which could be tricky to support in a more general way. There seem to be a few types of multi-agent: (shared reward, shared model), (shared reward, non-shared model), (non-shared reward, shared model), (non-shared-reward, non-shared model).

Seems like this PR can cover the first two cases. The third you could possibly handle by splitting the observations up so each agent's experience is handled separately. The fourth is the tricky one, the only easy thing that comes to mind is Philipp's suggestion to train one type of agent a time, so it can be treated as the third case.

Reply:
@ericl 
A quick question, I'm currently doing the splitting of the inputs and outputs in a new multiagent model; I'm passing in a set of lists indicating how the splits should be done. I'm doing this because the model doesn't have access to the original environment observation and action space. This doesn't seem like a good way, do you have any suggestions?

Reply:
Yeah, in the Carla env code, we make the original space dimensions a configuration parameter in catalog.py. You can do the same (make the original obs spaces a required configuration parameter of the model).

There might be a nicer way but this should be fine as long as we document it well.

Reply:
Submitted a new pull request that I think is a much cleaner approach so I'll close this one then?